### DISCUSSION TITLE
HIP 336 - Approval and Allowance API for Tokens

### DISCUSSION THREAD

**Discussion by superboo** [hiero-ledger/hiero-improvement-proposals#337] [Edits]
Discuss HIP-336 - Approval and Allowance API for Tokens 

[#336](https://github.com/hashgraph/hedera-improvement-proposal/pull/336)

**Comment by bugbytesinc** [comment]
Why is it necessary to specify `Owner` when it is redundant because it can only be the `Payer/operator` of the transaction?

https://github.com/hashgraph/hedera-protobufs/blob/77685bd9ef73302feff040a4c05605c863adecb0/services/response_code.proto#L1205

**Reply by mohsin-hedera** [reply]
Thanks for pointing this out. This has since been deprecated.

**Comment by hashaxis** [comment]
As a marketplace, token allowances are a superior solution to an escrow method.

However:

“Each account is limited to 100 allowances. This limit spans hbar and fungible token allowances and non-fungible token approved_for_all grants. There is no limit on the number of NFT serial number approvals an owner may grant.”

The arbitrary limit of 100 allowances creates a similar issue for a marketplace which the arbitrary limit of token associations created:

Which is that we’d need to create many unnecessary accounts just to scale our ability to list tokens if implementing this.

The limit on token associations was removed which allowed us to scale without creating unnecessary accounts. We would suggest that this 100 allowance limit is scaled to at least 10,000 or removed entirely if feasible.

**Reply by HGraphPunks** [reply]
I agree that this restriction will hurt non-escrow NFT marketplaces. Having the ability to facilitate 10s of thousands of token allowances will be necessary for marketplaces as the HBAR NFT scene gains more and more adoption.

**Reply by Ashe-Oro** [reply]
just to make sure I understand correctly. This limitation is on the user who is approving the grant, right? For example: I have 101 NFT Collections, I would only be able to set approvals for 100 of them? 

This does not affect the account which receives the approval? For example: As a NFT marketplace owner, my account can receive approvals for >100 token_ids.   

Or are both limited to 100 allowances (both give and receive)?

**Reply by hashaxis** [reply]
@Ashe-Oro 

"There is no limit on the number of NFT serial number approvals an owner may grant.” -- Give is unlimited.

"Each account is limited to 100 allowances." -- Receive is limited to 100.

**Reply by Ashe-Oro** [reply]
NFT serial number = 0.0.1043046/1 (unlimited)
NFT collection = 0.0.1043046 (limited to 100 approvals granted/received?)
FT = 0.0.1045728 (limited to 100 approvals granted/received?)

**Reply by Ashe-Oro** [reply]
also what about given approval to a smart contract? Is it impacted at all by this limitation. If I find out, I'll post here.

**Reply by hashaxis** [reply]
It would seem the allowance limit is set on an **account** (not on the collection token), but this is an assumption. Unclear how smart contract allowances are impacted if at all.

**Reply by hashaxis** [reply]
@Ashe-Oro has confirmed that the assumption that an allowance is set at the account level for the receiver was incorrect, and that this limit actually applies to the grantor. Thank you!
